parl/framework/algorithm.py

```diff
@@ -0,0 +1,188 @@
+
+
+class Model(Network):
+    """
+    A Model is owned by an Algorithm. It implements all the network model of
+    a specific problem.
+    """
+
+    def __init__(self):
+        super(Model, self).__init__()
+
+    def get_input_specs(self):
+        """
+        Output: list of tuples
+        """
+        raise NotImplementedError()
+
+    def get_state_specs(self):
+        """
+        States are optional to a Model.
+        Output: list of tuples
+        """
+        return []
+
+    def get_action_specs(self):
+        """
+        Output: list of tuples
+        """
+        raise NotImplementedError()
+
+    def get_reward_specs(self):
+        """
+        By default, a scalar reward.
+        User can specify a vector of rewards for some problems
+        """
+        return [("reward", dict(shape=[1]))]
+
+    def policy(self, inputs, states):
+        """
+        Return: action_dists: a dict of action distribution objects
+                states
+                An action distribution object can be created with
+                PolicyDistribution().
+        Optional: a model might not always have to implement policy()
+        """
+        raise NotImplementedError()
+
+    def value(self, inputs, states):
+        """
+        Return: values: a dict of estimated values for the current observations and states
+                        For example, "q_value" and "v_value"
+        Optional: a model might not always have to implement value()
+        """
+        raise NotImplementedError()
+
+
+class Algorithm(object):
+    """
+    An Algorithm implements two functions:
+    1. predict() computes forward
+    2. learn() computes a cost for optimization
+
+    An algorithm should be only part of a network. The user only needs to
+    implement the rest of the network in the Model class.
+    """
+
+    def __init__(self, model, hyperparas, gpu_id):
+        assert isinstance(model, Model)
+        check_duplicate_spec_names(model)
+        self.model = model
+        self.hp = hyperparas
+        self.gpu_id = gpu_id
+
+    def get_input_specs(self):
+        return self.model.get_input_specs()
+
+    def get_state_specs(self):
+        return self.model.get_state_specs()
+
+    def get_action_specs(self):
+        """
+        For non-RL algortihms, this can return []
+        """
+        return self.model.get_action_specs()
+
+    def get_reward_specs(self):
+        """
+        For non-RL algortihms, this can return []
+        """
+        return self.model.get_reward_specs()
+
+    def before_every_batch(self):
+        """
+        A callback function inserted before every batch of training.
+        See ComputationTask.learn()
+        """
+        pass
+
+    def after_every_batch(self):
+        """
+        A callback function inserted after every batch of training.
+        See ComputationTask.learn()
+        """
+        pass
+
+    def predict(self, inputs, states):
```

Comment:
abstractmethod? To be same as skylian did in https://github.com/PaddlePaddle/PARL/pull/12